### PR TITLE
Improve node error handling

### PR DESCRIPTION
**Pull request by ivov**
## Introduction

This PR implements and expands on [the 2020 MLH Fellowship's proposal on node error handling](https://hackmd.io/@jh6EYijgQ-GeAm9l3y-a4g/rksOxZkfv). This proposal standardizes access to the **status code** and **error message** in the API response error object on a per-node basis, 

## Original Proposal

Instead of sequential access checks for properties in the API response error object...

```ts
if (error.response && error.response.body && error.response.body.message) {
    // Try to return the error prettier
    throw new Error(`Spotify error response [${error.statusCode}]: ${error.response.body.message}`);
}
```

The original proposal maps the paths to them:

```ts
const box2ToN8nErrorAPITransformer: N8nErrorAPI = {
    code: ["error", "error", "status"],
    message: ["error", "error", "message"]
};
```

And handles/rethrows the error from a centralized location:

```ts
// GenericFunctions.ts (node-specific)
const errorObj = errorHandler(error, box2ToN8nErrorAPITransformer);
userDisplayNodeError("Spotify", errorObj.code, errorObj.message);

// centralized location
const recursivelyFindAction = (accumulator, currentValue) => accumulator[currentValue];

/**
* Takes in any API's error response object and uses the provided API to n8n Error Transformer to generate an n8n standard formatted error object.
* @param errorObj {any} - The API's Error Response Object
* @param pathTransformer {N8nErrorAPI} - The paths to parse in the errorObj to fetch the correct values.
* @return {N8nErrorAPIResponse} - The n8n standard API error object
*/
export const errorHandler = (errorObj: object, pathTransformer: N8nErrorAPI): N8nErrorAPIResponse => {
  const apiStandard: N8nErrorAPIResponse = {code: '', message: ''};

  for (const [key, value] of Object.entries(pathTransformer)) {
    apiStandard[key] = value.reduce(recursivelyFindAction, errorObj).toString();
  }

  return apiStandard;
};

const userDisplayNodeError = (nodeName: string, statusCode: string, errorMessage: string): string => {
	throw new Error(`${nodeName} error response [${statusCode}]: ${errorMessage}`);
};
```

The property access syntax logic is similar to [Lodash's _.get](https://lodash.com/docs/4.17.15#get)).

## Expanded Proposal

In this implementation, a number of changes were introduced to the original proposal.

Functions and interfaces are renamed more explicitly:

```ts
export interface IN8nErrorPathMapping {
	[key: string]: string[];
	code: string[];
	message: string[];
}

export interface IN8nApiResponseError {
	[key: string]: string;
	code: string;
	message: string;
}
```

Error handling logic is refactored as a single operation; some typings are tightened. Note the `never` type should be used for [functions that do not return](https://www.typescriptlang.org/docs/handbook/basic-types.html#never). 

```ts
/**
 * Standardizes an API response error and throws it.
 *
 * @export
 * @param {string} nodeName
 * @param {object} errorObject
 * @param {IN8nErrorPathMapping} errorPathMapping
 * @returns {never}
 */
export function handleError(
	nodeName: string,
	errorObject: object,
	errorPathMapping: IN8nErrorPathMapping
): never {
	const apiResponseError = standardizeError(errorObject, errorPathMapping);
	throwApiResponseError(nodeName, apiResponseError);
}

/**
 * Converts an API error object into a standard N8N error object based on the error path mapping provided.
 *
 * @export
 * @param {object} errorObject
 * @param {IN8nErrorPathMapping} errorPathMapping
 * @returns {IN8nApiResponseError}
 */
function standardizeError(errorObject: object, errorPathMapping: IN8nErrorPathMapping): IN8nApiResponseError {
	const apiResponseError: IN8nApiResponseError = {
		code: '',
		message: '',
	};

	const findValueRecursively = (accumulator: any, currentValue: any) => accumulator[currentValue];

	Object.entries(errorPathMapping).forEach(([key, path]) => {
		apiResponseError[key] = path.reduce(findValueRecursively, errorObject).toString();
	});

	return apiResponseError;
};

/**
 * Throws an error with a standard N8N error message.
 *
 * @export
 * @param {string} nodeName
 * @param {IN8nApiResponseError} apiResponseError
 * @returns {never}
 */
function throwApiResponseError(
	nodeName: string,
	apiResponseError: IN8nApiResponseError,
): never {
	throw new Error(`${nodeName} error response [${apiResponseError.code}]: ${apiResponseError.message}`);
};
```

End result at call site:

```ts
try {
	return await this.helpers.request!(options);
} catch (error) {

	const errorPathMapping: IN8nErrorPathMapping = {
		code: ["error", "status"],
		message: ["error", "error"],
	};

	handleError("Hacker News", error, errorPathMapping);
}
```

## Implementation

The PR implements the expanded proposal in...
- the Hacker News node,
- the Spotify node, and
- the G Suite Admin node.

In all three cases, the new error-handling style **worked well** and was relatively quick to implement. Identifying the paths and verifying that they are correct is the only slightly time-consuming step.

image

spotify_error

image

## Limitations

This proposal works well for those nodes that rely on a long chain of sequential access checks—a majority of all nodes. However, other error-handling patterns, such as those below, are not natively compatible with the proposal.

| Compatible? | Amount | Percentage |
|-------------|--------|------------|
| Yes         | 3      | 2%         |
| Likely      | 98     | 66%        |
| Maybe       | 29     | 20%        |
| Unlikely    | 3      | 2%         |
| Unknown     | 15     | 10%        |

_For the full breakdown, see the [next comment](https://github.com/n8n-io/n8n/pull/1309#issuecomment-755560062)._

We could look into whether it is possible to adjust the proposal for these other cases:
- multiple error paths possible,
- multiple errors joined together,
- non-REST API implementations, and
- miscellaneous error-handling variants.

### Multiple error paths possible

Nodes: HubSpot, Toggl, etc.

```ts
} catch (error) {
	if (error.statusCode === 403) {
		throw new Error('The Toggle credentials are probably invalid!');
	}

	const errorMessage = error.response.body && (error.response.body.message || error.response.body.Message);
	if (errorMessage !== undefined) {
		throw new Error(errorMessage);
	}

	throw error;
}
```

### Multiple errors joined together

Nodes: Strapi, Strava, Twake, etc.

```ts
} catch (error) {
	if (error.response && error.response.body && error.response.body.message) {

		let messages = error.response.body.message;

		if (Array.isArray(error.response.body.message)) {
			messages = messages[0].messages.map((e: IDataObject) => e.message).join('|');
		}
		// Try to return the error prettier
		throw new Error(
			`Strapi error response [${error.statusCode}]: ${messages}`,
		);
	}
	throw error;
}
```

### Non-REST API implementations

Nodes: Slack, Humantic AI, SMS77, etc.

```ts
	if (response.ok === false) {
		throw new Error('Slack error response: ' + JSON.stringify(response));
	}

	return response;
} catch (error) {
	if (error.statusCode === 401) {
		// Return a clear error
		throw new Error('The Slack credentials are not valid!');
	}

	if (error.response && error.response.body && error.response.body.message) {
		// Try to return the error prettier
		throw new Error(`Slack error response [${error.statusCode}]: ${error.response.body.message}`);
	}

	// If that data does not exist for some reason return the actual error
	throw error;
}
```

### Miscellaneous error-handling variants

Nodes: PayPal, SalesForce, Shopify, etc.

```ts
} catch (error) {

	if (error.response.body) {
		let errorMessage = error.response.body.message;
		if (error.response.body.details) {
			errorMessage += ` - Details: ${JSON.stringify(error.response.body.details)}`;
		}
		throw new Error(errorMessage);
	}

	throw error;
}
```

## Open Questions

In order of importance:

- Is it possible to adjust the error-handling proposal to cover the "Maybe" and "Unknown" cases?
- Where should we place `IN8nErrorPathMapping` and `IN8nApiResponseError`? The path in this PR `packages/workflow/src/Interfaces.ts` is just a provisional location.
- Where should we place the error handling functionality? The path in this PR `packages/core/src/NodeExecutionFunctions.ts` is just a provisional location.
- Is there a way to "auto-discover" the API response error object's structure so as to auto-generate the error path mapping?
- How should we properly type the parameters in `findValueRecursively`?

### CODE REVIEW

**Comment by ivov** [comment]
Error handling across the 149 nodes in the codebase that use a `request()` method.

Utility nodes (e.g. HTML Extract) and non-request nodes (e.g. Kafka) excluded.

**Compatibility key:**
- Yes: Implemented and operation verified.
- Likely: Error handling through sequential access check.
- Maybe: With adjustments, the proposal might possibly work.
- Unlikely: Even with adjustments, the proposal seems unlikely to work.
- Unknown: The error is just rethrown, so the API docs need to be revisited.

| Node                           | Compatible? | Reason if not compatible        |
|--------------------------------|-------------|---------------------------------|
| Google G Suite Admin           | Yes         |                -                |
| Hacker News                    | Yes         |                -                |
| Spotify                        | Yes         |                -                |
| Affinity                       | Likely      |                -                |
| Airtable                       | Likely      |                -                |
| Automizy                       | Likely      |                -                |
| AWS                            | Likely      |                -                |
| Bannerbear                     | Likely      |                -                |
| Bitbucket                      | Likely      |                -                |
| Bitly                          | Likely      |                -                |
| Brandfetch                     | Likely      |                -                |
| Calendly                       | Likely      |                -                |
| CircleCI                       | Likely      |                -                |
| Clearbit                       | Likely      |                -                |
| ClickUp                        | Likely      |                -                |
| Clickify                       | Likely      |                -                |
| Cockpit                        | Likely      |                -                |
| Coda                           | Likely      |                -                |
| Contentful                     | Likely      |                -                |
| ConvertKit                     | Likely      |                -                |
| Copper                         | Likely      |                -                |
| Cortex                         | Likely      |                -                |
| Customer.io                    | Likely      |                -                |
| Disqus                         | Likely      |                -                |
| Drift                          | Likely      |                -                |
| Dropbox                        | Likely      |                -                |
| Egoi                           | Likely      |                -                |
| EventBrite                     | Likely      |                -                |
| Facebook                       | Likely      |                -                |
| Flow                           | Likely      |                -                |
| FreshDesk                      | Likely      |                -                |
| GetResponse                    | Likely      |                -                |
| Ghost                          | Likely      |                -                |
| GitHub                         | Likely      |                -                |
| GitLab                         | Likely      |                -                |
| Google Sheet                   | Likely      |                -                |
| Google Translate               | Likely      |                -                |
| Gotify                         | Likely      |                -                |
| Gumroad                        | Likely      |                -                |
| Harvest                        | Likely      |                -                |
| HelpScout                      | Likely      |                -                |
| Intercom                       | Likely      |                -                |
| Iterable                       | Likely      |                -                |
| Jira                           | Likely      |                -                |
| Keap                           | Likely      |                -                |
| Line                           | Likely      |                -                |
| LingvaNex                      | Likely      |                -                |
| LinkedIn                       | Likely      |                -                |
| MailChimp                      | Likely      |                -                |
| MailerLite                     | Likely      |                -                |
| Mandrill                       | Likely      |                -                |
| Matrix                         | Likely      |                -                |
| Mattermost                     | Likely      |                -                |
| Mautic                         | Likely      |                -                |
| MessageBird                    | Likely      |                -                |
| Microsoft Excel                | Likely      |                -                |
| Microsoft OneDrive             | Likely      |                -                |
| Microsoft Outlook              | Likely      |                -                |
| Microsoft Teams                | Likely      |                -                |
| Mocean                         | Likely      |                -                |
| MondayCom                      | Likely      |                -                |
| MSG91                          | Likely      |                -                |
| NASA                           | Likely      |                -                |
| NextCloud                      | Likely      |                -                |
| OpenThesaurus                  | Likely      |                -                |
| Orbit                          | Likely      |                -                |
| Paddle                         | Likely      |                -                |
| PagerDuty                      | Likely      |                -                |
| Phantombuster                  | Likely      |                -                |
| Philips Hue                    | Likely      |                -                |
| PipeDrive                      | Likely      |                -                |
| Postmark                       | Likely      |                -                |
| ProfitWell                     | Likely      |                -                |
| Pushbullet                     | Likely      |                -                |
| Pushcut                        | Likely      |                -                |
| Quickbase                      | Likely      |                -                |
| RocketChat                     | Likely      |                -                |
| SignI4                         | Likely      |                -                |
| Spontit                        | Likely      |                -                |
| StoryBlok                      | Likely      |                -                |
| Stripe                         | Likely      |                -                |
| SurveyMonkey                   | Likely      |                -                |
| Telegram                       | Likely      |                -                |
| TheHive                        | Likely      |                -                |
| Todoist                        | Likely      |                -                |
| TravisCI                       | Likely      |                -                |
| Twilio                         | Likely      |                -                |
| Twist                          | Likely      |                -                |
| Typeform                       | Likely      |                -                |
| Unleashed Software             | Likely      |                -                |
| Uproc                          | Likely      |                -                |
| Vero                           | Likely      |                -                |
| Webflow                        | Likely      |                -                |
| Wekan                          | Likely      |                -                |
| WooCommerce                    | Likely      |                -                |
| WordPress                      | Likely      |                -                |
| Wufoo                          | Likely      |                -                |
| Yourls                         | Likely      |                -                |
| Zendesk                        | Likely      |                -                |
| Zoho                           | Likely      |                -                |
| Zulip                          | Likely      |                -                |
| HubSpot                        | Maybe       | Multiple error paths            |
| Toggl                          | Maybe       | Multiple error paths            |
| Mindee                         | Maybe       | Multiple errors joined together |
| Pushover                       | Maybe       | Multiple errors joined together |
| Sendy                          | Maybe       | Multiple errors joined together |
| Strapi                         | Maybe       | Multiple errors joined together |
| Strava                         | Maybe       | Multiple errors joined together |
| Twake                          | Maybe       | Multiple errors joined together |
| Twitter                        | Maybe       | Multiple errors joined together |
| Vonage                         | Maybe       | Multiple errors joined together |
| Xero                           | Maybe       | Multiple errors joined together |
| Asana                          | Maybe       | Multiple errors joined together |
| Box                            | Maybe       | Multiple errors joined together |
| Google Analytics               | Maybe       | Multiple errors joined together |
| Google Books                   | Maybe       | Multiple errors joined together |
| Google Calendar                | Maybe       | Multiple errors joined together |
| Google Cloud Natural Language  | Maybe       | Multiple errors joined together |
| Google Contacts                | Maybe       | Multiple errors joined together |
| Google Drive                   | Maybe       | Multiple errors joined together |
| Google Firebase CloudFirestore | Maybe       | Multiple errors joined together |
| Google Firebase RealtimeDB     | Maybe       | Multiple errors joined together |
| Google Gmail                   | Maybe       | Multiple errors joined together |
| Google Task                    | Maybe       | Multiple errors joined together |
| Google YouTube                 | Maybe       | Multiple errors joined together |
| Invoice Ninja                  | Maybe       | Multiple errors joined together |
| S3                             | Maybe       | Miscellaneous variant           |
| SalesForce                     | Maybe       | Miscellaneous variant           |
| PayPal                         | Maybe       | Miscellaneous variant           |
| Shopify                        | Maybe       | Miscellaneous variant           |
| Humantic AI                    | Unlikely    | Not REST                        |
| Slack                          | Unlikely    | Not REST                        |
| SMS77                          | Unlikely    | Not REST                        |
| Hunter                         | Unknown     | No error handling               |
| ActiveCampaign                 | Unknown     | No error handling               |
| Acuity Scheduling              | Unknown     | No error handling               |
| AgileCRM                       | Unknown     | No error handling               |
| CoinGecko                      | Unknown     | No error handling               |
| FileMaker                      | Unknown     | No error handling               |
| JotForm                        | Unknown     | No error handling               |
| Medium                         | Unknown     | No error handling               |
| Salesmate                      | Unknown     | No error handling               |
| Segment                        | Unknown     | No error handling               |
| Sentry.io                      | Unknown     | No error handling               |
| Taiga                          | Unknown     | No error handling               |
| Trello                         | Unknown     | No error handling               |
| Upload                         | Unknown     | No error handling               |
| Zoom                           | Unknown     | No error handling               |

**Review by krynble — CHANGES_REQUESTED** [review]
Great job so far! I like the concept, this will be very helpful.

I had a few issues with existing nodes that were not changed. Can we improve the way we display errors for nodes that weren't changed?

We have some generic `throw new Error()` in the code and they won't display nicely in these cases. See the attached screenshot below. You can see the thrown error on the left side panel.

Generic throw error

I have seen this behavior on other nodes like Clickup and Airtable. Either we simply display the error message or nothing at all.

Clickup node error

This is related to the `Unknown` nodes in the compatibility table.

File: packages/workflow/src/NodeErrors.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+		switch (this.httpCode.charAt(0)) {
+			case '4':
+				this.message = STATUS_CODE_MESSAGES['4XX'];
+				break;
+			case '5':
+				this.message = STATUS_CODE_MESSAGES['5XX'];
+				break;
+			default:
+				this.message = UNKNOWN_ERROR_MESSAGE;
```
[the remark this answers is not in the document]
Comment (BHesseldieck) [reply packages/workflow/src/NodeErrors.ts:223]: agree, we just need to identify the error code first. Need to check what are the most common keys in the nodes that use error-codes 👍🏼

**Comment by Colton** [comment]
Been having similar issues with trying to implement this kind of error handling for the QuickBase node. This PR looks great! Can't wait to test it out. Thank you @ivov !

**Comment by ivov** [comment]
Thanks! @BHesseldieck contributed heavily to this as well.

**Comment by janober** [comment]
Amazing work! Got merged and will be released with the next version.

**Comment by janober** [comment]
Got released with n8n@0.116.0

**Pull request merged**